tests/zero-downtime-scaling/zero-downtime-scaling.md

```diff
@@ -0,0 +1,381 @@
+# Zero-Downtime Scaling
+
+This document describes a test plan for testing zero-downtime scaling of NGF.
+
+*Zero-downtime scaling* means that when the number of replicas of NGF is scaled up or down, the clients don't experience
+any interruptions to the traffic they send to applications exposed via NGF.
+
+<!-- TOC -->
+- [Zero-Downtime Scaling](#zero-downtime-scaling)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Test Environment](#test-environment)
+  - [Tests](#tests)
+    - [Setup](#setup)
+    - [Scale Gradually](#scale-gradually)
+    - [Scale Abruptly](#scale-abruptly)
+    - [Analyze](#analyze)
+<!-- TOC -->
+
+## Goals
+
+- Ensure that scaling up NGF doesn't lead to any loss of traffic flowing through the data plane.
+- Ensure that scaling down NGF doesn't lead to any loss of traffic flowing through the data plane.
+- Ensure that after scaling, NGF can process changes to resources.
+
+## Non-Goals
+
+- Testing long-lived connections.
+
+  When scaling the number of NGF replicas down, Kubernetes will shut down existing NGF pods by sending a SIGTERM. If the
+  pod doesn't terminate in 30 seconds (the default period), Kubernetes will send a SIGKILL.
+
+  When proxying Websocket or any long-lived connections, NGINX will not terminate until that connection is closed by
+  either the client or the backend. This means that unless all those connections are closed by clients/backends before
+  or during an upgrade (which is highly unlikely), NGINX will not terminate, which means Kubernetes will kill NGINX. As
+  a result, the clients will see the connections abruptly closed and thus experience downtime.
+
+  As a result, we *will not* use any long-live connections in this test, because NGF cannot support zero-downtime
+  upgrades in this case.
+
+- Scaling to over 25 replicas.
+
+## Test Environment
+
+The scaling tests will be performed on two different Kubernetes clusters:
+
+- A Kubernetes cluster with 10 nodes on GKE:
+  - Node: n2d-standard-4 (4 vCPU, 8GB memory).
+  - Enabled GKE logging.
+- A Kubernetes cluster with 25 nodes on GKE:
+  - Node: n2d-standard-4 (4 vCPU, 8GB memory).
+  - Enabled GKE logging.
+
+The 10 Node cluster will cover the case where multiple NGF Pods are running on the same Node, whereas the 25 Node
+cluster will test 1 NGF Pod per Node.
+
+The rest of the test environment is the same for both cases:
+
+- Tester VMs:
```

Comment:
The scale script requires jq. Perhaps it make sense to add another machine with configured kubectl and jq to the environment?

Reply:
Good callout. I think I'll remove `jq` from the script since it's not required -- it just makes the output prettier. 

Running the script from a VM would be obnoxious because the script and the manifests would have to be copied to the VM.